### PR TITLE
Bump top-level @types/node version

### PR DESCRIPTION
**Pull request by weswigham**
Followup to #77784, since the same error happens upon `yarn compile` once `yarn` is fixed due to a slightly out of date `@types/node` dependency version in the vscode root~

With this change, _every_ `package.json` inside the vscode repo has an `@types/node` version of `^10.14.8`. It'd be nice if they could somehow be kept in sync~

### CODE REVIEW

**Comment by weswigham** [comment]
Neato, apparently bumping the top-level node version breaks a bunch.

**Comment by weswigham** [comment]
@rbuckton I get some impossible errors in this PR:
```
[15:57:30] Error: E:/Github/vscode/node_modules/@types/node/ts3.2/util.d.ts(6,15): Cannot redeclare block-scoped variable 'custom'.
[15:57:30] Error: E:/Github/vscode/node_modules/@types/node/ts3.2/util.d.ts(9,15): Cannot redeclare block-scoped variable 'custom'.
[15:57:30] Error: E:/Github/vscode/node_modules/@types/node/index.d.ts(82,15): Cannot redeclare block-scoped variable 'custom'.
[15:57:30] Error: E:/Github/vscode/node_modules/@types/node/index.d.ts(85,15): Cannot redeclare block-scoped variable 'custom'.
```
It shouldn't be possible for the `3.2+` and older `@types/node` entrypoints to both be in the compilation, and yet, there they are, conflicting. I am unsure about the best way to continue from here, hmmm

File: src/vs/base/common/async.ts [hunk 1]
```diff
@@ -35,6 +35,7 @@ export function createCancelablePromise<T>(callback: (token: CancellationToken)
 	});
 
 	return new class implements CancelablePromise<T> {
+		[Symbol.toStringTag]!: string;
```
Comment (jrieken) [inline src/vs/base/common/async.ts]: We still support IE and there is no support for `Symbol`

File: src/typings/es6-promise.d.ts [hunk 2]
```diff
@@ -8,75 +8,77 @@ interface Thenable<T> {
...
+	 */
+	then<U>(onFulfilled?: (value: T) => U | Thenable<U>, onRejected?: (error: any) => U | Thenable<U>): Promise<U>;
+	then<U>(onFulfilled?: (value: T) => U | Thenable<U>, onRejected?: (error: any) => void): Promise<U>;
+
+	/**
+	 * Sugar for promise.then(undefined, onRejected)
+	 *
+	 * @param onRejected called when/if "promise" rejects
+	 */
+	catch<U>(onRejected?: (error: any) => U | Thenable<U>): Promise<U>;
```
Comment (jrieken) [inline src/typings/es6-promise.d.ts]: After all those changes, does the copy right header still makes sense or is this a new version of said file?

File: src/vs/workbench/api/browser/mainThreadWorkspace.ts [hunk 3]
```diff
@@ -123,11 +123,11 @@ export class MainThreadWorkspace implements MainThreadWorkspaceShape {
 
 	// --- search ---
 
-	$startFileSearch(includePattern: string | null, _includeFolder: UriComponents | null, excludePatternOrDisregardExcludes: string | false | null, maxResults: number | null, token: CancellationToken): Promise<UriComponents[] | null> {
+	$startFileSearch(includePattern: string | undefined, _includeFolder: UriComponents | undefined, excludePatternOrDisregardExcludes: string | false | undefined, maxResults: number | undefined, token: CancellationToken): Promise<UriComponents[] | undefined> {
```
Comment (jrieken) [inline src/vs/workbench/api/browser/mainThreadWorkspace.ts]: Cosmetic changes? There is debate in the team if `undefined` or `null` should be used and I don't wanna end up bike shedding. I prefer `undefined` but I will not accept a PR with such changes unless a new version of TS mandates it

File: src/vs/workbench/contrib/extensions/browser/extensions.contribution.ts [hunk 4]
```diff
@@ -278,6 +278,7 @@ CommandsRegistry.registerCommand({
 		} catch (e) {
 			onUnexpectedError(e);
 		}
+		return;
```
Comment (jrieken) [inline src/vs/workbench/contrib/extensions/browser/extensions.contribution.ts]: ?

**Review by jrieken — CHANGES_REQUESTED** [review]

File: src/vs/base/common/async.ts (hunk 1 above)
Comment (weswigham) [reply src/vs/base/common/async.ts]: Without an initializer, this is (sortof) a type only position - it makes the checker happy when node pulls in the lib with symbol, but doesn't actually affect emit.

File: src/vs/workbench/api/browser/mainThreadWorkspace.ts (hunk 3 above)
Comment (weswigham) [reply src/vs/workbench/api/browser/mainThreadWorkspace.ts]: Uhhh, some of these APIs flip-flopped weather they used `null` or `undefined` while I was writing the PR (vscode master was briefly broken, even) and I just did my best to reconcile at each step. IIRC this one I swapped to stifle a mismatch triggered by a change made in the vscode `master` branch somewhere else.

File: src/vs/workbench/contrib/extensions/browser/extensions.contribution.ts (hunk 4 above)
Comment (weswigham) [reply src/vs/workbench/contrib/extensions/browser/extensions.contribution.ts]: All these async functions reported `Not all code paths return a value` once I bumped the node version. I just added returns to squeltch the error. I've no idea why it only started being reported then, honestly.

File: src/typings/es6-promise.d.ts (hunk 2 above)
Comment (weswigham) [reply src/typings/es6-promise.d.ts]: Eh? I just took the class-based definitions and unwrapped them into interface based ones so they'd merge correctly with the lib promise, if present. It's all the same types and signatures, just in a slightly adjusted set of wrapping constructs.

**Comment by jrieken** [comment]
> inside the vscode repo has an @types/node version of ^10.14.8. It'd be nice if they could somehow be kept in sync

Problem is that we are actually only shipping version `10.11.0` (the one that comes with Electron 4.2.x). The fact that we declare version 10.12.x seems already fishy to me. Assigning @bpasero who seems to have been doing this in the past

**Comment by rbuckton** [comment]
@weswigham: The errors you are getting seem to be due to how vscode's build is set up. I've been investigating but don't have a solution quiet yet. Here are some of my findings so far:

In *build/lib/compilation.ts* they use gulp-tsb with `noFileSystemLookup = true`. The LanguageServiceHost created by gulp-tsb doesn't read files from the file system when `noFileSystemLookup = true`, so when we try to look up the package.json for node to read the `"typesVersions"` field, we fail to find it. Setting `noFileSystemLookup = false` doesn't seem to be the solution, however.

I also tried excluding `node_modules/@types/node/index.d.ts` from `typesDts` in *build/lib/compilation.ts* which gets rid of the duplicate identifier errors, however it results in new errors in that many of the files in *src/typings/* produce a "Cannot find the type definition file for 'node'" error. 

@jrieken: Given that this seems to partly be an issue with gulp-tsb, could you provide any insight as to what might be needed to address this?

**Comment by bpasero** [comment]
I think this is a good start, but I would suggest to do this when we update to Electron 6 where we get node.js 12 as part of it. 

I do not think that types/node is following the same versioning scheme as node.js, so it will probably be hard to get the exact same d.ts file that matches the version we are using in Electron. 

I still see compile errors btw when trying this out:

image

**Comment by jrieken** [comment]
> @jrieken: Given that this seems to partly be an issue with gulp-tsb, could you provide any insight as to what might be needed to address this?

The idea is that all files must come in via `gulp.src` and not be fetched/added later. Usually that's files in src and types. Seems like @mjbvz added more support for node_modules/@types (see https://github.com/microsoft/vscode/blob/c8b0433eabc17b900b10156bda646f3a1aa2bc72/build/lib/compilation.ts#L84) and that might be the way how things come in twice

**Comment by weswigham** [comment]
From what @rbuckton mentioned to me the other day, that way of including @types packages is flawed, as it doesn't include any of the `package.json` entries (which include such stylings as where the package entrypoint is and what files to include for which versions of TS) and always includes every declaration file in the package, meaning errors like the above can occur, where types intended for two different versions of TS are both loaded. And since .json files aren't valid root files, you can't exactly pass them in that list to get them injected into the virtual filesystem.

**Comment by rbuckton** [comment]
@weswigham is correct. We determine which folder to load types from by reading a `"typesVersions"` field in package.json, which isn't available to gulp-tsb when `noFileSystemLookup` is `true`.

I experimented with setting that field to `false`, but ran into other issues due to collisions with `xterm` and two other `xterm`-related packages that ship their own types and are *also* defined in `src/typings`.

**Comment by rbuckton** [comment]
@jrieken also note that if I explicitly exclude `!node_modules/@types/node/index.d.ts` there are other errors that crop up from package definitions in `src/typings` that use `/// <reference types="node" />` since they can't find the `index.d.ts` for Node and there's no `package.json` they can look at to find the correct types.

**Comment by bpasero** [comment]
Since this seems to need a considerable amount of TS wisdom, I am assigning @mjbvz and leave it independent from the Electron 6 update. I think it is good to merge this either way, even with the latest 10.x bindings.

**Comment by rbuckton** [comment]
I've created https://github.com/jrieken/gulp-tsb/pull/81 to address the `gulp-tsb` build issue. Once that is merged, we would need to make the following changes:

- Include `'node_modules/@types/*/package.json'` in `typesDts` so that the package.json files are known to `gulp-tsb` when `noFilesystemLookup` is `true`.
- Set `excludeNodeModulesFromRootNames` to `true` on the options passed to `tsb.create`

**Comment by mjbvz** [comment]
I used a [terrible monkey patch](https://github.com/microsoft/vscode/commit/cf446b92337b2dca065bd9a95ecfb4314c4b8a9f#diff-61376a469d8efd2e0a8472412929c72aR83) to update us to TS 3.6 without updating the `@types/node` version.

We'll see if we can switch over to standard TSC for builds: #80074

**Comment by mjbvz** [comment]
Closing as we've fixed the typesVersion support in gulp-tsb and are building VS Code on TS 3.6

**Pull request closed without merging**